### PR TITLE
Feature/rail-split-implementation

### PR DESCRIPTION
**Pull request by PiterGroot**

### CODE REVIEW

File: Assets/Scenes/Develop.unity
Comment (Joost-Robat) [inline Assets/Scenes/Develop.unity:1]: Don't forget to revert the changes of the Engine health and the Ignore coal burn bool on the Oven.

File: Assets/Prefabs/Train/Tracks/TrainTracks.prefab
Comment (Joost-Robat) [inline Assets/Prefabs/Train/Tracks/TrainTracks.prefab:1]: Perhaps change this to TrainTracksPart1 to be more clear.

File: Assets/Prefabs/Train/Tracks/TrainTracks1.prefab
Comment (Joost-Robat) [inline Assets/Prefabs/Train/Tracks/TrainTracks1.prefab:1]: Perhaps change this to TrainTracksPart2 to be more clear.

File: Assets/Prefabs/Train/Tracks/TrainTracks2.prefab
Comment (Joost-Robat) [inline Assets/Prefabs/Train/Tracks/TrainTracks2.prefab:1]: Perhaps change this to TrainTracksPart3 to be more clear.

File: Assets/Scripts/Framework/DamageRepairManagement/Damageables/BoxDamageable.cs [hunk 1]
```diff
@@ -1,6 +1,8 @@
 using FishNet.Object;
 using UnityEngine;
 
+using DerailedDeliveries.Framework.Train;
+
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/DamageRepairManagement/Damageables/BoxDamageable.cs]:
```suggestion
```

File: Assets/Scripts/Framework/Train/BadRailSplitDamager.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    /// Class responsible for damaging all boxes inside the train which 
+    /// are not stored in a shelf when train is riding on a bad split.
+    /// </summary>
+    [RequireComponent(typeof(TrainController))]
+    public class BadRailSplitDamager : NetworkBehaviour
+    {
+        private TrainController _trainController;
+
+        [SerializeField]
+        private float _damageInterval;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Train/BadRailSplitDamager.cs]:
```suggestion
        [SerializeField]
        private float _damageInterval;

        private TrainController _trainController;
```

File: Assets/Scripts/Framework/Train/TrainController.cs [hunk 3]
```diff
@@ -209,7 +264,51 @@ private void SwitchCurrentTrack(int trackID, bool setDistanceAlongSpline = false
...
+                    bool badSplitDirection = BadRailSplitOrder[clampedIndex];
+
+                    // Check if train wants to go to a bad rail split.
+                    if (badSplitDirection == TrainEngine.Instance.CurrentSplitDirection)
+                        IsOnBadRailSplit = true;
+                }
+                else
+                {
+                    IsOnBadRailSplit = false;
+                }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Train/TrainController.cs:288]:
```suggestion
                else
                    IsOnBadRailSplit = false;
```

**Review by Joost-Robat — COMMENTED** [review]
Looks good, minor comments.

File: Assets/Scripts/Framework/Camera/CameraShaker.cs [hunk 4]
```diff
@@ -32,9 +58,19 @@ private void Awake()
...
+
             _multiChannelPerlin.m_AmplitudeGain = 0;
         }
 
+        private void HandleRailSplitChanged(bool badRailSplit)
+        {
+            float newCameraFrequency = badRailSplit ? _badRailSplitShakeFrequencyPenalty : _startCameraNoiseFrequency;
+
+            DOTween.To(() => _multiChannelPerlin.m_FrequencyGain, x 
+                => _multiChannelPerlin.m_FrequencyGain = x, newCameraFrequency, 1);
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Camera/CameraShaker.cs]: Should the `1` be a `SerializeField`?

File: Assets/Scripts/Framework/DamageRepairManagement/Damageables/BoxDamageable.cs [hunk 5]
```diff
@@ -31,6 +26,11 @@ private protected override void TakeDamage()
             base.TakeDamage();
         }
 
+        /// <summary>
+        /// Method to force this <see cref="BoxDamageable"/> to take damage externally.
+        /// </summary>
+        public void ForceTakeDamage() => TakeDamage();
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/DamageRepairManagement/Damageables/BoxDamageable.cs]: You can also just make `TakeDamage()` public.

File: Assets/Scripts/Framework/Train/BadRailSplitDamager.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+namespace DerailedDeliveries.Framework.Train
+{
+    /// <summary>
+    /// Class responsible for damaging all boxes inside the train which 
+    /// are not stored in a shelf when train is riding on a bad split.
+    /// </summary>
+    [RequireComponent(typeof(TrainController))]
+    public class BadRailSplitDamager : NetworkBehaviour
+    {
+        private TrainController _trainController;
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Train/BadRailSplitDamager.cs]: Move a bit down.

File: Assets/Scripts/Framework/Train/BadRailSplitDamager.cs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+        [SerializeField]
+        private float _damageInterval;
+
+        private BoxDamageable[] boxDamageablesInTrain;
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Train/BadRailSplitDamager.cs]:
```suggestion
        private BoxDamageable[] _boxDamageablesInTrain;
```

File: Assets/Scripts/Framework/Train/BadRailSplitDamager.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+        private bool _isTrainMoving;
+
+        private protected float p_damageIntervalElapsed;
+
+        private void Awake() => _trainController = gameObject.GetComponent<TrainController>();
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Train/BadRailSplitDamager.cs]:
```suggestion
        private void Awake() => _trainController = GetComponent<TrainController>();
```

File: Assets/Scripts/Framework/Train/BadRailSplitDamager.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            if (TrainEngine.Instance == null)
+                return;
+
+            TrainEngine.Instance.OnSpeedChanged -= OnVelocityChanged;
+        }
+
+        [Server]
+        private void OnVelocityChanged(float velocity) => _isTrainMoving = Mathf.Abs(velocity) > 0.1f;
+
+        private void HandleRailSplitChanged(bool badSplit)
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Train/BadRailSplitDamager.cs]:
```suggestion
        private void HandleRailSplitChanged(bool isBadSplit)
```

File: Assets/Scripts/Framework/Train/BadRailSplitDamager.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            if (!badSplit)
+            {
+                boxDamageablesInTrain = null;
+                return;
+            }
+
+            boxDamageablesInTrain = FindObjectsOfType<BoxDamageable>()
+                .Where(boxDamageable => boxDamageable.IsInTrain).ToArray();
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Train/BadRailSplitDamager.cs]: You can just get all TrainDamageables and damage them. This way the shelf also gets damage.

File: Assets/Scripts/Framework/Train/TrainController.cs [hunk 11]
```diff
@@ -75,6 +90,12 @@ public float DistanceAlongSpline
         /// </summary>
         public TrainEngine TrainEngine { get; private set; }
 
+        /// <summary>
+        /// Invokes when train switches from rail split and returns if train is on bad split.
+        /// <br/> bool == false = bad side of split.
+        /// </summary>
+        public Action<bool> onRailSplitChange;
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/Train/TrainController.cs]:
```suggestion
        public Action<bool> OnRailSplitChange;
```

**Review by lucasvdiepen — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Framework/Train/TrainController.cs (hunk 3 above)
Comment (PiterGroot) [reply Assets/Scripts/Framework/Train/TrainController.cs:288]: If either the if or the else in the statement contains multiple lines both the if and the else need brackets.  Here in the [code convention](https://github.com/lucasvdiepen/DerailedDeliveries/wiki/Code-Conventions#if-statements).

Comment (Joost-Robat) [reply Assets/Scripts/Framework/Train/TrainController.cs:288]: You are correct, I wasn't paying attention!

**Review by Joost-Robat — COMMENTED** [review]
Looks good, when updated with develop has to be checked again though.

**Review by Joost-Robat — APPROVED** [review]
Looks good, works good, very good! Ready for merge.

File: Assets/Art/2D/Textures/Environment/T_SplitsingLeft.png
Comment (lucasvdiepen) [inline Assets/Art/2D/Textures/Environment/T_SplitsingLeft.png:1]: `Splitsing` should be translated to english. It is also missing a suffix.

File: Assets/Art/2D/Textures/Environment/T_SplitsingRight.png
Comment (lucasvdiepen) [inline Assets/Art/2D/Textures/Environment/T_SplitsingRight.png:1]: `Splitsing` should be translated to english. It is also missing a suffix.

File: Assets/Art/Materials/Environment/M_SplitsingLeft.mat
Comment (lucasvdiepen) [inline Assets/Art/Materials/Environment/M_SplitsingLeft.mat:1]: `Splitsing` should be translated to english.

File: Assets/Art/Materials/Environment/M_SplitsingRight.mat
Comment (lucasvdiepen) [inline Assets/Art/Materials/Environment/M_SplitsingRight.mat:1]: `Splitsing` should be translated to english.

File: Assets/Scripts/Framework/DamageRepairManagement/Damageable.cs [hunk 12]
```diff
@@ -49,7 +49,7 @@ public class Damageable : NetworkBehaviour
         public virtual void Repair() => ChangeHealth(MaxHealth);
 
         [Server]
-        private protected virtual void TakeDamage()
+        public virtual void TakeDamage()
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/DamageRepairManagement/Damageable.cs:55]: This needs a summary now that it is public.

File: Assets/Scripts/Framework/DamageRepairManagement/Damageables/BoxDamageable.cs [hunk 13]
```diff
@@ -29,7 +29,7 @@ private protected override void UpdateTimer()
         }
 
         [Server]
-        private protected override void TakeDamage()
+        public override void TakeDamage()
```
Comment (lucasvdiepen) [inline Assets/Scripts/Framework/DamageRepairManagement/Damageables/BoxDamageable.cs:35]: This is missing a summary but can use `<inheritdoc>`.

**Review by lucasvdiepen — CHANGES_REQUESTED** [review]

**Review by lucasvdiepen — APPROVED** [review]

**Pull request merged**